nncf/common/graph/graph.py

```diff
@@ -241,17 +366,218 @@ def add_edge(self, u_of_edge: str, v_of_edge: str, **attrs):
+                attrs_node['label'] = node['label']
+            if 'style' in node:
+                attrs_node['style'] = node['style']
+
+            out_graph.add_node(node_name, **attrs_node)
+        if extended:
+            for u, v in self._nx_graph.edges:
+                out_graph.add_edge(u, v, label=self._nx_graph.edges[u, v][NNCFGraph.ACTIVATION_SHAPE_EDGE_ATTR])
+        else:
+            for u, v in self._nx_graph.edges:
+                out_graph.add_edge(u, v)
+
+        return out_graph
+
+    def _get_graph_for_visualization(self) -> nx.DiGraph:
+        """A user-friendly graph .dot file, making it easier to debug the network and setup
+        ignored/target scopes."""
+        out_graph = nx.DiGraph()
+        for node_name, node in self._nx_graph.nodes.items():
+            attrs_node = {}
+            attrs_node['label'] = str(node[NNCFGraph.ID_NODE_ATTR]) + ' ' + str(NNCFGraph.NODE_NAME_ATTR)
+            out_graph.add_node(node_name, **attrs_node)
+
+        for u, v in self._nx_graph.edges:
+            out_graph.add_edge(u, v, label=self._nx_graph.edges[u, v][NNCFGraph.ACTIVATION_SHAPE_EDGE_ATTR])
+
+        mapping = {k: v["label"] for k, v in out_graph.nodes.items()}
+        out_graph = nx.relabel_nodes(out_graph, mapping)
+        for node in out_graph.nodes.values():
+            node.pop("label")
+
+        return out_graph
+
+    def get_node_by_name(self, name: NNCFNodeName) -> NNCFNode:
+        matches = []
+        for nx_node in self._nx_graph.nodes.values():
+            if nx_node[NNCFGraph.NODE_NAME_ATTR] == name:
+                matches.append(nx_node)
+        if not matches:
+            raise RuntimeError("Could not find a node {} in NNCFGraph!".format(name))
+        if len(matches) > 1:
+            raise RuntimeError("More than one node in NNCFGraph matches name {}:\n{}".
+                               format(name,
+                                      '\t\n'.join(
+                                          [n[NNCFGraph.KEY_NODE_ATTR] for n in matches])))
+        return self._nx_node_to_nncf_node(next(iter(matches)))
+
+    def __eq__(self, other: 'NNCFGraph'):
+        nm = iso.categorical_node_match([NNCFGraph.ID_NODE_ATTR,
+                                         NNCFGraph.KEY_NODE_ATTR,
+                                         NNCFGraph.MODULE_ATTRIBUTES], [None, None, None])
+        em = iso.categorical_edge_match([NNCFGraph.ACTIVATION_SHAPE_EDGE_ATTR,
+                                         NNCFGraph.IN_PORT_NAME_EDGE_ATTR], [None, None])
+        return nx.is_isomorphic(self._nx_graph, other._nx_graph, node_match=nm, edge_match=em)
+
+    def is_output_node(self, node: NNCFNode) -> bool:
+        return not list(self._nx_graph.successors(self._node_id_to_key_dict[node.node_id]))
+
+    def get_nx_graph_copy(self) -> nx.DiGraph:
+        return deepcopy(self._nx_graph)
+
+    def get_matching_nncf_graph_pattern_io_list(self, expression: Expression) -> List[NNCFGraphPatternIO]:
```

Comment:
It looks like NNCFGraph became a 'monster' class. What do you think about separating algorithms that work with the graph into separate entities? I mean the pattern matching algorithm and traversal algorithms.

Reply:
Good idea, will try to do this.

Reply:
Will implement in a separate PR, created a ticket 57305 for this